### ISSUE TITLE
use of abtract base classes, methods and properties

### ISSUE THREAD

**Issue by gschramm** [deepinv/deepinv#373]
Hi all,

have you considered using python's abstract base classes for your abstract base classes (e.g. Physics or LinearPhysics)?

The advantage would be that people subclassing e.g. LinearPhysics would exacatly know which methods have to be implemented and which methods are inhereted (by default).

E.g. sth like:
```python
import abc

class Physics(nn.Module, ABC):
    def __init__(self):
        super().__init__()
   
    @property
    @abstractmethod
    def my_abstract_property(self):
        # a property that must be implemented by the subclass
        pass
 
    @abstractmethod
    def forward(self, x, **kwargs):
        # an abstract methods that needs to be defined by the subclass
        pass

class LinearPhysics(Physics):
    # still an abstract class since not all abstract methods have been implemented
    # on top we define a new abstract A_A_adjoint method
    def __init__(self):
        super().__init__()

    @abstractmethod
    def A_A_adjoint(self, y, **kwargs):
        pass
```

The principle advantage is that people trying to subclass e.g. `LinearPhysics` but do not implement all abstract methods,
will immediately get an error when trying to create an object of the new subclass.

**Comment by Andrewwango** [comment]
Thanks for the interesting idea @gschramm, I guess this is primarily to aid contributors rather than users, to make sure you've implemented the right methods?

I guess it doesn't have any disadvantages, except for adding extra lines which may put off new contributors coming from a pure Python, rather than say, a C#, background.

**Comment by gschramm** [comment]
Exactly, this is primarily for contributers (or users that just want to implement / modify physics).
I think the overhead is minimal, but the benefits are big.

I think the concept is using abstract base classes is not language specific and well supported in python and C++.

It also makes testing whether a specific derived class is "valid" much easiert.

**Comment by tachella** [comment]
very nice suggestion, we should add this to the TODO list!

**Comment by jscanvic** [comment]
Hi Georg, I believe it's a very worthwhile change but it looks tricky to implement it for `Physics` and its derivatives, unlike for say `Loss`, `Transform`, `ImageDataset`, etc.

The main issue with `Physics` is that it somehow manages to never raise an error when you don't implement things because it relies on attributes passed in at construction as the fallback, e.g., `self.forw`. The other base classes on the other hand directly raise a `NotImplementedError` so they're way closer to abstract classes already.

Are you happy with rescoping the issue to those base classes instead and maybe deal with `Physics` at a later stage in another issue?

**Comment by gschramm** [comment]
Sure. I wasn't aware of the construciton time behavior of `Physics`. As long as this is documented / clear for developers that is not a problem. Maybe at the end of the `__init__()` you could check whether the methods that you need are indeed present.
